### PR TITLE
BUG: Fix interp issues with small arrays

### LINKED ISSUE #5705: interpolate needs fixing for short arrays of sample points.
See #5594.

### PR DESCRIPTION
**Pull request by jaimefrio**
Fixes #5705

### CODE REVIEW

**Comment by jaimefrio** [comment]
I haven't been able to produce the segfault I predicted for small arrays, so not sure how to test this.

I have also cleaned up a little the code and got rid of some code duplication. I was expecting a minor performance slowdown, but it is actually a tad faster than current master after these changes. Using @newville's script from #5594 I get these timings:

```
 Numpy Version:           this PR    master
 much_finer_grid          0.11425    0.11600   
 finer_grid               0.02887    0.03325
 similar_grid             0.01120    0.01528
 coarser_grid             0.01332    0.01811
 much_coarser_grid        0.00837    0.01102
 finer_stretched_grid     0.02894    0.03872
 similar_stretched_grid   0.02892    0.03297
 finer_compressed_grid    0.02852    0.03072
 similar_compressed_grid  0.01101    0.01525
 warped_grid              0.01139    0.01411
 very_low_noise_grid      0.02986    0.03415
 low_noise_grid           0.03593    0.04406
 med_noise_grid           0.04654    0.05673
 high_noise_grid          0.09372    0.11115
 very_high_noise_grid     0.13594    0.15036
 extreme_noise_grid       0.22071    0.21804
 random_fine_grid         0.46477    0.46065
 random_grid              0.79051    0.82327
```

**Comment by seberg** [comment]
Maybe we can add a test that would show up in valgrind?

File: numpy/core/src/multiarray/compiled_base.c [hunk 1]
```diff
@@ -637,72 +671,67 @@ arr_interp(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwdict)
...
-        slopes = (double *) PyArray_malloc((lenxp - 1)*sizeof(double));
-        if (! slopes) {
-            goto fail;
-        }
+    /* binary_search_with_guess needs at least a 2 item long array */
+    if (lenxp == 1) {
+        const npy_double xp_val = dx[0];
+        const npy_double fp_val = dy[0];
+
         NPY_BEGIN_ALLOW_THREADS;
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Could probably make this key size dependent at some point.

File: numpy/core/src/multiarray/compiled_base.c [hunk 2]
```diff
@@ -637,72 +671,67 @@ arr_interp(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwdict)
...
+    }
+    else {
+        npy_intp j = 0;
 
-            if (npy_isnan(x)) {
-                dres[i] = x;
-                continue;
+        /* only pre-calculate slopes if there are relatively few of them. */
+        if (lenxp <= lenx) {
+            slopes = PyArray_malloc((lenxp-1) * sizeof(npy_double));
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Spaces around `-`.

File: numpy/core/src/multiarray/compiled_base.c [hunk 3]
```diff
@@ -637,72 +671,67 @@ arr_interp(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwdict)
...
             if (j == -1) {
                 dres[i] = lval;
             }
-            else if (j == lenxp - 1) {
-                dres[i] = dy[j];
-            }
             else if (j == lenxp) {
                 dres[i] = rval;
             }
+            else if (j == lenxp-1) {
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Spaces around `-`.

File: numpy/core/src/multiarray/compiled_base.c [hunk 4]
```diff
@@ -525,35 +528,63 @@ binary_search_with_guess(double key, double arr [], npy_intp len,
...
+        else {
+            /* key > arr[guess-1] */
+            return guess - 1;
+        }
     }
-    if ((guess < len - 9) && (key <= arr[guess + 8])) {
-        imax = guess + 8;
+    else {
+        /* key > arr[guess] */
+        if (key <= arr[guess+1]) {
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Spaces around `+`.

File: numpy/core/src/multiarray/compiled_base.c [hunk 5]
```diff
@@ -525,35 +528,63 @@ binary_search_with_guess(double key, double arr [], npy_intp len,
...
+            return guess;
+        }
+        else {
+            /* key > arr[guess+1] */
+            if (key <= arr[guess+2]) {
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Spaces around `+`.

**Comment by charris** [comment]
LGTM modulo style nitpicks.

**Comment by jaimefrio** [comment]
Pushed new commit.

**Comment by charris** [comment]
Merged, thanks Jaime.

**Pull request merged**